### PR TITLE
Publisher: Do not show help dialog when pressing enter

### LINKED ISSUE #5791: Bug: Pressing Enter in publisher unexpectedly pops up Help dialog
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior:

When pressing the Enter key when in a text field in the new publisher the Help dialog pops up.

publisher_help_dialog_on_enter

### Expected Behavior:

Help dialog should not appear.

### Version

3.17.3-nightly.2

### What platform you are running OpenPype on?

Windows

### Steps To Reproduce:

1. Open Publisher
2. Enter a textfield
3. Press enter

The help dialog appears.

Note: This cannot be reproduced in Maya somehow - likely Maya captures the enter key event instead.

I was able to reproduce this in **Tray Publisher, Houdini, MRV2 and Fusion** hosts - so I assume it's actually a bug and not specific to a certain DCC even though it can't be reproduced in Maya.

### Are there any labels you wish to add?

- [X] I have added the relevant labels to the bug report.

### Relevant log output:

_No response_

### Additional context:

Reported [here](https://github.com/ynput/OpenPype/pull/5744#pullrequestreview-1684904063) and [more details here](https://github.com/ynput/OpenPype/pull/5744#issuecomment-1768371129)

[cuID:[OP-7218](https://app.clickup.com/t/86bw9bv52)]

### PR DESCRIPTION
**Pull request by BigRoy**
## Changelog Description

Do not show help dialog when pressing enter; avoid Help button taking focus

## Additional info

Fix #5791

## Testing notes:

1. Pressing enter key in e.g. variant field to create instance should not pop-up help dialog

_Note that this already wasn't the case in Maya, but everywhere else it did occur. So in Maya it should show no difference in behavior._

### CODE REVIEW

**Review by LiborBatek — CHANGES_REQUESTED** [review]
Did tested in ```max 2023``` and it fixes the issue with showing up ```help``` pop up window but introduces strange switching to the other publish pane....see the vid for more insight.

Need to be addressed.

https://github.com/ynput/OpenPype/assets/112623825/103d529e-90a6-4e25-95cd-f03babbadad7

File: openpype/tools/publisher/widgets/help_widget.py [hunk 1]
```diff
@@ -12,6 +12,7 @@ class HelpButton(QtWidgets.QPushButton):
     def __init__(self, parent):
         super(HelpButton, self).__init__(parent)
         self.setObjectName("CreateDialogHelpButton")
+        self.setFocusPolicy(QtCore.Qt.NoFocus)
```
Comment (iLLiCiTiT) [inline openpype/tools/publisher/widgets/help_widget.py:15]: The object on it's own cannot modify focus. The window must be first shown to be able to define focus.

Comment (BigRoy) [reply openpype/tools/publisher/widgets/help_widget.py:15]: The logic works fine however - what do you mean? :)

This is the focus policy - not the active focus.

**Comment by BigRoy** [comment]
> Did tested in `max 2023` and it fixes the issue with showing up `help` pop up window but introduces strange switching to the other publish pane....see the vid for more insight.
> 
> Need to be addressed.
> Publisher_Enter_Confirm.mp4

@LiborBatek Ok, so this is due to the Help button not receiving focus but now the Create tab (top left) takes the focus and thus enter clicks that tab entry. When I disable the focus policy for the tabs themselves then the next focus is the "Save" icon bottom right. As such clicking enter "saves" the settings which I'm not sure is preferred either.

@iLLiCiTiT I feel like we're better off ensuring that the KeyPressEvents doesn't pass through to begin with from the Create Options Widgets and the Instance Publish Option widgets

Currently whenever actively being in any of the fields in this part of the UI:
image

And this one:
image

Actually applies the Enter key event to the focused object, like the Help button, the Tabs above it or the Controls below it. I'd actually argue however that if currently any of the entries in the screenshotted areas above are active that it should not pass through key events (or at least not enter keys?) We'd still want to be able to TAB out of it however?

**Review by LiborBatek — APPROVED** [review]
All good when hitting ```Enter``` key and no any unwanted changes happening!

image

**Comment by iLLiCiTiT** [comment]
Can we add the ignore logic to existing method in window?
```python
if event.key() in {
    QtCore.Qt.Key_Return,
    QtCore.Qt.Key_Enter,
}:
    event.accept()
    return
```

**Comment by BigRoy** [comment]
> Can we add the ignore logic to existing method in window?
> 
> ```python
> if event.key() in {
>     QtCore.Qt.Key_Return,
>     QtCore.Qt.Key_Enter,
> }:
>     event.accept()
>     return
> ```

As far as I know we can not put it on the Window itself - because children widgets as far as I know will still be first receiving the event before it hits the Windows' event upstream. We could install an event filter on the window which triggers for ALL children events but that'd just be slower / getting way more events than we'd like.

Basically the logic is this:

- text field, press enter. does the text field process the event?
- while not processed, does the parent process the event? (recursively going up the parents to the first one dealing with the event)

**Comment by iLLiCiTiT** [comment]
> As far as I know we can not put it on the Window itself - because children widgets as far as I know will still be first receiving the event before it hits the Windows' event upstream

I've tried and it does not trigger the button... I think the default is triggered only if the event is not accepted at any child of the window hierarchy.

**Comment by BigRoy** [comment]
@iLLiCiTiT seems to work - like this?

**Review by antirotor — COMMENTED** [review]
Weird - If you open for example max and publisher, hit enter, then help window will appear. But after reset, enter won't open it anymore. Same thing in Houdini

**Comment by BigRoy** [comment]
> Weird - If you open for example max and publisher, hit enter, then help window will appear. But after reset, enter won't open it anymore. Same thing in Houdini

That's because the active button then has changed to the reset button most likely (or some other field). Anyway, that goes to show that this _new fix_ with the event capture on the window does not work and we should likely revert again to this commit https://github.com/ynput/OpenPype/pull/5838/commits/46064849e1cca25cfc0d47f236902fe87f44511f

@antirotor can you confirm that commit instead works as intended?

**Comment by antirotor** [comment]
> @antirotor can you confirm that commit instead works as intended?

it doesn't work either - I replicated the original change in new code:

```python
    def keyPressEvent(self, event):
        # Ignore escape button to close window
        if event.key() == QtCore.Qt.Key_Escape:
            event.accept()
            return

        # Avoid enter key to activate the "help" dialog, fix #5791
        return super().keyPressEvent(event)
```

also it behaves differently when you click on different parts of the publisher window. On some parts, the help window won't pop-up, but on the others it will.

**Comment by BigRoy** [comment]
> > @antirotor can you confirm that commit instead works as intended?
> 
> it doesn't work either - I replicated the original change from [4606484](https://github.com/ynput/OpenPype/commit/46064849e1cca25cfc0d47f236902fe87f44511f) in the new code:
> 
> ```python
>     def keyPressEvent(self, event):
>         # Ignore escape button to close window
>         if event.key() == QtCore.Qt.Key_Escape:
>             event.accept()
>             return
> 
>         # Avoid enter key to activate the "help" dialog, fix #5791
>         return super().keyPressEvent(event)
> ```
> 
> also it behaves differently when you click on different parts of the publisher window. On some parts, the help window won't pop-up, but on the others it will.

Note that it was on a different class - not on the window. Did you do that? Best to just checkout that commit actually.

**Comment by antirotor** [comment]
> Note that it was on a different class - not on the window. Did you do that? Best to just checkout that commit actually.

Well, the original commit works, but not for the first time - when you open the Publisher and immediately hit enter, help will open. But once you click on anything inside or reset, enter will no longer open help window.

**Comment by BigRoy** [comment]
> > Note that it was on a different class - not on the window. Did you do that? Best to just checkout that commit actually.
> 
> Well, the original commit works, but not for the first time - when you open the Publisher and immediately hit enter, help will open. But once you click on anything inside or reset, enter will no longer open help window.

That sounds fine - maybe somewhat confusing, but more acceptable than pressing enter in the text fields doing it.

@iLLiCiTiT thoughts? Revert to that commit - other options?

**Comment by mkolar** [comment]
This issue was resolved by merging https://github.com/ynput/OpenPype/pull/6107

closing.

**Pull request closed without merging**